relay-server/src/actors/processor.rs

```diff
@@ -433,55 +435,150 @@ fn track_sampling_metrics(
+///
+/// Additionally, processing applies clock drift correction using the system clock of this Relay, if
+/// the Envelope specifies the [`sent_at`](Envelope::sent_at) header.
+#[derive(Debug)]
+pub struct ProcessMetrics {
+    /// A list of metric items.
+    pub items: Vec<Item>,
+
+    /// The target project.
+    pub project_key: ProjectKey,
+
+    /// The instant at which the request was received.
+    pub start_time: Instant,
+
+    /// The value of the Envelope's [`sent_at`](Envelope::sent_at) header for clock drift
+    /// correction.
+    pub sent_at: Option<DateTime<Utc>>,
+}
+
+/// Applies HTTP content encoding to an envelope's payload.
+///
+/// This message is a workaround for a single-threaded upstream actor.
+#[derive(Debug)]
+pub struct EncodeEnvelope {
+    request: SendEnvelope,
+}
+
+impl EncodeEnvelope {
+    /// Creates a new `EncodeEnvelope` message from `SendEnvelope` request.
+    pub fn new(request: SendEnvelope) -> Self {
+        Self { request }
+    }
+}
+
+/// CPU-intensive processing tasks for envelopes.
+#[derive(Debug)]
+pub enum EnvelopeProcessor {
+    ProcessEnvelope(Box<ProcessEnvelope>),
+    ProcessMetrics(Box<ProcessMetrics>),
+    EncodeEnvelope(Box<EncodeEnvelope>),
 }
 
 impl EnvelopeProcessor {
     pub fn from_registry() -> Addr<Self> {
         REGISTRY.get().unwrap().processor.clone()
     }
+}
 
-    /// Starts a multi-threaded envelope processor.
-    pub fn start(
-        config: Arc<Config>,
-        _redis: Option<RedisPool>,
-    ) -> Result<Addr<Self>, ServerError> {
-        let thread_count = config.cpu_concurrency();
-        relay_log::info!("starting {} envelope processing workers", thread_count);
+impl relay_system::Interface for EnvelopeProcessor {}
+
+impl FromMessage<ProcessEnvelope> for EnvelopeProcessor {
+    type Response = relay_system::NoResponse;
+
+    fn from_message(message: ProcessEnvelope, _sender: ()) -> Self {
+        Self::ProcessEnvelope(Box::new(message))
+    }
+}
 
+impl FromMessage<ProcessMetrics> for EnvelopeProcessor {
+    type Response = NoResponse;
+
+    fn from_message(message: ProcessMetrics, _: ()) -> Self {
+        Self::ProcessMetrics(Box::new(message))
+    }
+}
+
+impl FromMessage<EncodeEnvelope> for EnvelopeProcessor {
+    type Response = NoResponse;
+
+    fn from_message(message: EncodeEnvelope, _: ()) -> Self {
+        Self::EncodeEnvelope(Box::new(message))
+    }
+}
+
+/// Service implementing the [`EnvelopeProcessor`] interface.
+///
+/// This service handles messages in a worker pool with configurable concurrency.
+pub struct EnvelopeProcessorService {
+    config: Arc<Config>,
+    #[cfg(feature = "processing")]
+    rate_limiter: Option<RedisRateLimiter>,
+    #[cfg(feature = "processing")]
+    geoip_lookup: Option<GeoIpLookup>,
+}
+
+impl EnvelopeProcessorService {
+    /// Creates a multi-threaded envelope processor.
+    pub fn foo(config: Arc<Config>, _redis: Option<RedisPool>) -> Result<Self, ServerError> {
```

Comment:
whoops

Reply:
Thanks, good catch! Reminds me that I wanted to untangle that constructor mess somehow :)